### ISSUE TITLE
Suggestion: Extension methods

### ISSUE THREAD

**Issue by RyanCavanaugh** [Microsoft/TypeScript#9]
Allow a declarative way of adding members to an existing type's prototype

Example:
```TypeScript

class Shape {
 // ...
}

/* ...elsewhere...*/
extension class Shape { // Syntax??
    getArea() { return /* ... */; }
}

var x = new Shape();
console.log(x.getArea()); // OK
```

**Comment by oliverw** [comment]
Reading this discussion I still do not understand why extension methods can't be implemented just like the static C# extension methods.  I mean why couldn't the compiler could just emit invocation of the static method?

**Comment by RyanCavanaugh** [comment]
Because TypeScript is not C# and JavaScript is not the CLR.

Here's some code that people would very reasonably expect to work; none of these would be possible if we tried to extension methods with call site rewrites.
```ts
// Setup
class Rectangle {
  constructor(public width: number, public height: number) {}
}
// Do not be distracted by exact syntax here
extension function getArea(this s: Rectangle ) {
  return s.width * s.height;
}
// end setup

// Challenge course begins
interface HasArea {
  getArea(): number;
}

function fn1(x: HasArea) {
  console.log(x.getArea());
}
fn1(new Rectangle(10, 10));

var a: any = Math.random() > 0.5 ? new Rectangle(3, 3) : { getArea: function() { return 3; } };
console.log(a.getArea());

var s = new Rectangle(5, 5);
var f = s.getArea.bind(s);
console.log(f);

class Square {
  constructor(public size: number) { }
  getArea() { return this.size * this.size; }
}
var box: Square|Rectangle = /* ... anything ...*/
console.log(box.getArea());

box.getArea.call(box);

var boxes: HasArea[] = [new Square(5), new Rectangle(10, 10)];
boxes.forEach(b => {
  console.log(b.getArea());
});
```

The only plausible language in which we rewrote call sites would have the following restrictions:
* You can't use extension methods as first-class objects, even though that's a very common JavaScript pattern. Only immediate invocations are allowed
* Extension methods aren't used when checking assignability, so you can't use them to add new functionality that is used with any indirection. You can't use an extension method to 'fix' a type to match some interface, even though that's a primary use case
* Extension methods aren't visible when put in a union type, which everyone will assume is a compiler bug
* If the type of a variable accidentally becomes `any`, your code will break at runtime with an error you won't be able to diagnose via debugging the code itself. This will look like a compiler bug to most people
* You can't use TypeScript extension methods from JavaScript without knowing the special code to write

When you add up all those restrictions, the *only* real advantage `foo.extMethod()` has over `extMethod(foo)` is that there's an extra dot in the extension method call. Plus, it creates a giant trap by having extension method calls on `any` variables fail. That's not a good deal.

**Comment by zlumer** [comment]
@RyanCavanaugh this is an insanely detailed explanation, thank you! When this thread started, I was sure that extension methods are a very good idea, now I believe we're better of without them (at least for now).

**Comment by jbondc** [comment]
@RyanCavanaugh Have you looked into using a ~ proxy:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

Not sure where that proposal stands but fairly optimistic this can work:
```javascript
class Rectangle {
  constructor(public width: number, public height: number) {}
  static foo() {
      return 'foo';
   }
}
Object.seal(Rectangle);
extends(Rectangle, {
  static bar(){  return 'bar'; }
  getArea() {
  return this.width * this.height;
});
```
Emits
```javascript
class Rectangle {
....
}
proxy_Rectangle = new Proxy(Rectangle,  {
    get: function(target, name){
        var extension = {
           bar: function() { return 'bar'; }
        } 
        return  name in extension ? extension[name] : target[name]
    }
});

proxy_Rectangle_new_handler =  {
    get: function(target, name){
        var extension = {
           getArea: function() { return this.width * this.height; }
        } 
        return  name in extension ? extension[name] : target[name]
    }
});

// All futur references to 'Rectangle' are wrapped by the proxy
function fn1(x: HasArea) {
  console.log(x.getArea());
}
fn1(new Proxy(new Rectangle(10, 10), proxy_Rectangle_new_handler));
proxy_Rectangle.bar()
```

**Comment by niemyjski** [comment]
+1

**Comment by jasonwilliams200OK** [comment]
At present, the primitive types can be extended as:

```typescript
// Extending String prototype
interface String {
    format(...params: any[]): string;
}

// using `any` so consumer can pass number or string,
// might be a better way to constraint types to
// string and number only using generic?
String.prototype.format = function (...params: any[]) {
    var s = this,
        i = params.length;

    while (i--) {
        s = s.replace(new RegExp('\\{' + i + '\\}', 'gm'), params[i]);
    }

    return s;
};

// Usage:
// 'Added {1} by {0} to your collection'.format(artist, title)
```

Would be nice to have `extension` keyword fit in that kind of context as well. :)

**Comment by roelvanlisdonk** [comment]
+1

**Comment by Kavignon** [comment]
How could someone contribute for the extension classes in Typescript ?

**Comment by bchavez** [comment]
Would be great if we could extend enums with this. Something like

```
enum View{
   Edit,
   Detail
}

View.Edit.asString()
```

vs what we have to do now:

```
View[View.Edit]
```

**Comment by RyanCavanaugh** [comment]
Note that you can add to an `enum` using `module` (or `namespace`):
```ts
enum E {
	A, B
}
module E {
	export function toString(e: E): string {
		return E[e];
	}
}

var x = E.toString(E.A);
```

**Comment by benliddicott** [comment]
#3508 looks like the compromise solution. Avoids backwardness, doesn't mess with prototypes and doesn't rely on type information. The only downside is it has a funny-looking syntax. I'd prefer `->` for `call`.

So this:

    o.my()::Extension.method().rocks();
    // or preferably this
    o.my()->Extension.method().rocks();

desugars to:

    Extension.method.prototype.call(o.my()).rocks();

There's a possible ambiguity if there were extension properties though, or methods which return extension methods.

**Comment by Zorgatone** [comment]
:+1:

**Comment by WanderWang** [comment]
It seems implement in TypeScript 1.7 

```
class Shape {
 // ...
}

interface Shape {
    getArea()
}


var x = new Shape();
console.log(x.getArea()); // OK

```

Should we close this issue ?

**Comment by Zorgatone** [comment]
That does not create a new method

**Comment by roelvanlisdonk** [comment]
Well this is exactly what I was looking for, so for me a YES, on closing this issue.

**Comment by WanderWang** [comment]
@Zorgatone 

you can create a new method with this way

```
class Shape {
 // ...
}

interface Shape {
    getArea()
}

Shape.prototype.getArea = function(){
}


var x = new Shape();
console.log(x.getArea()); // OK

```

**Comment by Zorgatone** [comment]
Oh, that's with prototype. I wasn't considering that.

I tried to do something similar, but without the interface trick :+1:

**Comment by niemyjski** [comment]
@WanderWang @Zorgatone You can't do that, I tried doing that in separate files using the es6 module syntax and got tons of errors. I think it only works with the legacy external modules.

**Comment by Zorgatone** [comment]
Uhm then keep the issue open :D

**Comment by falsandtru** [comment]
-1 Humans cannot understand code flow using extension methods. Extension should be declarative feature.

**Comment by electricessence** [comment]
The problem with doing this is that (as hopefully we’ve all dealt with), is it could pollute the existing code in a more complex environment.
“Extensions” would simply be compiler sugar for writing utility methods that would pass the ‘this’ along.

**Comment by jasonwilliams200OK** [comment]
> -1 Humans cannot understand code flow using extension methods. 

(-1'ing everything is not really helping)

There are countless real world use-cases where extension methods are proven to be useful.

If you are unsure about the significance of extension methods, take a look at C#.

If you don't like the way it is proposed to be implemented, please provide an alternate approach.

> Extension should be declarative feature.

Yes and this is in the very first line of this proposal.

**Comment by falsandtru** [comment]
I don't think that it is declarative because we cannot get the whole story. We must find pieces of a class.

**Comment by jasonwilliams200OK** [comment]
If you are building the library yourself, then you should be able to add the additional methods in the class. If you are working with someone's else library (which is most likely documented) then you can add extension methods elsewhere in your code without touching `vendor/**` code. Finding the pieces to form the 'whole story' is a non-issue especially when editors are there to assist you.

Splitting the code into separate files vs. dumping [everything in one place](https://github.com/cesanta/v7/blob/master/v7.c), both have pros and cons. Giving this choice to the developer however they want to consume the feature (or avoid it) would be the best strategy from TypeScript team's standpoint. Any feature of any language can be misused.

**Comment by coderfox** [comment]
:+1:

**Comment by coderfox** [comment]
:+1: 

I would suggest that Extension is only useful for telling the compiler there is an existing member for the existing class, so the code:

```ts
class A {}
extension class A {
  func() : void =>  {
  }
}
```

could be just compiled to:

```js
class A{}
A.prototype.func = () : void => {
}
```

but such Extension is of great use to intellisense.

**Comment by ronzeidman** [comment]
+1
I think the extension methods should support interfaces and be a syntactical sugar like in C#:
```typescript
interface IMovable {
    position: Point;
}
extension class MyExtensions {
    move(this moveable: IMovable, newPoint: Point) {
...
    }
}
var car: Movable = {...};
car.move(newPoint);
```
this will compile to:
```javascript
var MyExtensions = {
      move...
}
var car = {...};
MyExtensions.move(car, newPoint);
```

**Comment by TrabacchinLuigi** [comment]
if this isn't already the case i think  @ronzeidman is absolutely right :+1:

**Comment by SaschaNaz** [comment]
If we are really hesitant to support `extension class Homu` syntax, can we just add type information from `Homu.prototype.method1 = () => {}`, similar to what Salsa does (#4955)?

**Comment by RyanCavanaugh** [comment]
@ronzeidman @TrabacchinLuigi you need to read https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592

**Comment by ronzeidman** [comment]
@RyanCavanaugh The example in https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592 is a good one. I've now read the entire thread and there were great suggestions, I like the ES7 bind operator (::) but I suggest something a bit different -
introducing an "Elixir" like pipe operator "|>":
```typescript
function getArea(x: Rectangle) { ... }

var myRectangle = new Rectangle();
myRectangle |> getArea();
```
it will make things like "extending" array functions much easier and much more readable without changing the prototype:
```typescript
grouppedReduce(groupBy(mySort(myArray.map(...), ...).filter(...), ...), ...);
myArray.map(...)|>mySort(...).filter(...)|>groupBy(...)|>grouppedReduce(...);
```

What do you say?

**Comment by RyanCavanaugh** [comment]
We're almost certainly not taking up any new runtime operators that don't have at least some traction with the ES committee (especially when there are competing operators with similar meaning), so taking that to ESDiscuss would be the next step if you'd like to see it happen.

**Comment by TrabacchinLuigi** [comment]
@RyanCavanaugh I like the bind operator too!

**Comment by ronzeidman** [comment]
@RyanCavanaugh I agree, and the bind operator (of which I guess people agree on since its an ES7 suggestion) could work really well, especially if you could configure the type of "this" the function can get like in https://github.com/Microsoft/TypeScript/issues/6018 suggestion

**Comment by hdachev** [comment]
Hey just want to express my +1 for @ronzeidman -style extension classes with a few more words. I fully realize that this is something that will differentiate TS from ES, since it's impossible to pull it off without a type system, but I also think that this is the sort of thing that leverages the type system for an amazing gain in developer experience, and it makes perfect sense for our language.

ES will obviously never get a type system which means it will never be able to get this feature either, while we can easily have and love it every day. It transpiles to ok looking JS, and to anyone thinking otherwise, well today I have to type a lot of the same stuff manually on a daily basis. Re: worries that methods won't behave as usual when deferenced or tested for existence, etc, I really don't see this a major problem, since that is rather niche usage wrt method calling. As such those can result in a compiler error and I just don't see how that can become a problem.

Aside from the obvious case for interfaces, it'd allow one to extend collection types like `MyType[]` that just cannot be done otherwise sensibly. Underscore-like utility libraries will become a syntactic bliss. I imagine it'll play very nicely with this-typing for building fluent apis too.

Brief - life will be greater, and more beautiful, and so much more fantastic! We have a type system, there's nothing wrong in using it to make our lives even better. This feature relies entirely on the type system, and as such, in my opinion, is a perfect fit for TS.

**Comment by gabrielPeart** [comment]
I believe the Extensions implemented in [Swift 2.1](https://developer.apple.com/library/ios/documentation/Swift/Conceptual/Swift_Programming_Language/Extensions.html) is a very good example of how it could be implemented (as initial proposal at least)

Will allow us to do something like the following example [here](https://gist.github.com/gabrielPeart/c1f61f1ec88e1da734b4)

**Comment by katsgeorgeek** [comment]
Class extension would be useful for separated definition members with different access level:

```javascript
class A {}

private extension class A {
    func1() : void =>  {}
}

protected extension class A {
    func2() : void =>  {}
}
```

**Comment by mattleibow** [comment]
Just wanted to give my humble suggestion, pardon my two-week TypeScript skills.

I have been reading this discussion and maybe we are trying to do something primarily designed for type-based systems. 
JavaScript is much less typey, so work with that (this syntax is just one way):


**Given this class:**

```
class Square {
    width: number;
    height: number;
}
```

**Case 1: Does not touch the prototype:**

```
extension class SimpleSquareExtensions {
    depth: number = 0;
    static scale: number = 0;

    // must be static
    static function getArea(this square: Square) {
        // Can't access `depth` at all, not static
        // Can't do this: `square.scale` - not on Square
        // Can do this: `SimpleSquareExtensions.scale`
        return square.width * square.height;
    }
}

// JS:
SimpleSquareExtensions.scale = 0;
SimpleSquareExtensions.prototype.depth = 0;
SimpleSquareExtensions.getArea = function (square) {
    return square.width * square.height;
}
```

This is a very basic type that does not touch the `Square` prototype, so:

```
var sq: Square = new Square();

var area = sq.getArea();
// JS:
// var area = SimpleSquareExtensions.getArea(sq);

var anySq: any = sq;
var anyArea = anySq.getArea(); // No Intellisense for getArea()
// JS:
// var anyArea = anySq.getArea();
// Runtime error - expected
```

This works like C# dynamic object, I can't apply a `Square` extension 
method to an object that might not be a square.

Also, we could, or not, permit other non-extension members on the class, such as 
the `depth` and `scale` properties. (could be methods too) 

**Case 2: Adds to the prototype:**

```
extension class MoreSquareExtensions {
    extension static scale: number = 0;
    extension depth: number = 0;

    extension static function getArea(this square: Square) {
        // Can't do this: `square.depth` - static method
        // Can't do this: `Square.depth` - not static field
        // Can't do this: `MoreSquareExtensions.depth` - not on MoreSquareExtensions
        // Can't do this: `square.scale` - static field
        // Can't do this: `MoreSquareExtensions.scale` - not on MoreSquareExtensions
        // Can do this: `Square.scale`
        return square.width * square.height;
    }

    // no need for static
    extension function getVolume(this square: Square) {   
        // Can't do this: `Square.depth` - not static field
        // Can't do this: `MoreSquareExtensions.depth` - not on MoreSquareExtensions
        // Can't do this: `square.scale` - static field
        // Can't do this: `MoreSquareExtensions.scale` - not on MoreSquareExtensions
        // Can do this: `square.depth`
        // Can do this: `Square.scale`             
        return square.width * square.height * square.depth;
    }
}

// JS:
Square.scale = 0;
Square.prototype.depth = 0;
Square.getArea = function (square) {
    return square.width * square.height;
}
Square.prototype.getVolume = function () {
    return this.width * this.height * this.depth;
}
```

Due to the `extension` keyword, the developer is instructing the compiler to
modify the type. The dev is responsible for resolving conflicts:

```
var sq: Square = new Square();

var area = sq.getArea();
// JS:
// var area = Square.getArea(sq);

var anySq: any = sq;
var anyArea = anySq.getArea(); // Still no Intellisense?
// JS:
// var anySq = sq;
// var anyArea = anySq.getArea();
// works as expected, because it is on the prototype

sq.depth = 10;
var volume = sq.getVolume();
// JS:
// sq.depth = 10;
// var volume = sq.getVolume();
```

**Comment by hdachev** [comment]
Case 1 would be the best thing since sliced bread because it'll work with
interfaces, generics, builtins and third party types extremely well. It's
an extremely modular feature, and will work automatically with future
extensions of the type system.

Case 2 only works for extending classes, can't work on interfaces, can't
work on builtins without polluting them, and can't work only on
select generic type parametrizations.

For example - in Case 1 - when imported, your getArea method will be
available on an HTMLCanvas element, because it has width and height
properties. How awesome is that?

**Comment by RyanCavanaugh** [comment]
@mattleibow seems like you're proposing partial classes, see #563

**Comment by frogcjn** [comment]
can we extension interface? interface can also have some implemented functions, just like Swift:
```typescript
interface Rect {
    x: number
    y: number
}

extension Rect {
    area() => x * y
}
```
Swift version:
```swift

protocol Rect {
    var x: Float {get}
    var y: Float {get}
}
extension Rect {
    func area()-> Float { x * y }
}
```

**Comment by SaschaNaz** [comment]
>can we extension interface?

It can be dangerous: https://github.com/Microsoft/TypeScript/issues/9#issuecomment-52329918, https://github.com/Microsoft/TypeScript/issues/9#issuecomment-52347966

**Comment by frogcjn** [comment]
@SaschaNaz I don't think so. Swift already deal with it.

**Comment by frogcjn** [comment]
@SaschaNaz I think interface extension is no more dangerous than class extension.

The way to solve the conflicting problem is using developer‘s import to decides which implementation should use

In Swift, when different module extends some global interface or properties, they always use perfix to identify the extension. For example, RxSwift package extends UIView class with rx_dellocated. When developer import "RxSwift", the UIView will have rx_deallocated property. If developer don't import "RxSwift", there is no rx_deallocated with UIView.

**Comment by RyanCavanaugh** [comment]
Given the constraints imposed by emit (see https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592), there are realistically only two paths forward for object-first non-property method invocations:

* Partial classes (#563)
* ES20xx 'bind' operator (#3508)

I think we're reasonably satisfied (minus bikeshedding) on partial classes, and these can already be emulated today with a not-distasteful syntax (elucidated here: https://github.com/Microsoft/TypeScript/issues/563#issuecomment-218016147). I'd encourage anyone to weigh in meaningfully on `extension` vs `partial` vs other keywords, *keeping in mind* that we didn't just say "yea `partial` sounds good" and be done with it.

The ES20xx 'bind' operator ('xx' because who knows what version) is still in semantic flux and we need that proposal to either die cleanly (to let us lay clear claim to that syntactic space), or proceed with clear semantics, before we can act on it. It's an area where TypeScript can't safely act unilaterally. There's good discussion going on at https://github.com/tc39/proposal-bind-operator/issues/24 and its parent repo that outlines next steps for the proposal and it'd be great for people to perform citizen advocacy at ESDiscuss or other forums to help this move forward.

**Comment by alonbardavid** [comment]
I'm not sure if this has been suggested yet (this is a long thread), but what about compiler warning or compiler errors with explicit override (like eslint "ignore this" comments) for call site replacement?

That is in the situation implied in this [comment](https://github.com/Microsoft/TypeScript/issues/9#issuecomment-52347966) , instead of silently failing in the second example an error is emitted akin to "extension method called on unknown type" and the following code can be used to remove the error:

```
process(((n) => {
    n.getArea(); //no extension
}));
```

I haven't seen this type of explicit compiler hints in typescript, and perhaps it's intentional, but I think that call site replacement is a very powerful tool that a majority of developers would love to use, and the situation described will almost never happen.
I don't think Typescript should fall into undefined behaviour at runtime, but I think that an error and an explicit optout will prevent anyone from unwittingly making this mistake.

**Comment by RyanCavanaugh** [comment]
@alonbardavid see https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592 for why call site rewriting really, truly, really just does not work.

**Comment by KingDavid12** [comment]
@RyanCavanaugh To take your example code from https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592 I don't entirely see why some amount of call site rewriting can't work at all. For instance could we do something like: 

```typescript
// Setup
class Rectangle {
  constructor(public width: number, public height: number) {}
}
// Do not be distracted by exact syntax here
extension function getArea(this s: Rectangle ) {
  return s.width * s.height;
}
// end setup

// Challenge course begins
interface HasArea {
  getArea(): number;
}

function fn1(x: HasArea) {
  console.log(x.getArea());
}
fn1(new Rectangle(10, 10));

var a: any = Math.random() > 0.5 ? new Rectangle(3, 3) : { getArea: function() { return 3; } };
console.log(a.getArea());

var s = new Rectangle(5, 5);
var f = s.getArea.bind(s);
console.log(f);

class Square {
  constructor(public size: number) { }
  getArea() { return this.size * this.size; }
}
var box: Square|Rectangle = /* ... anything ...*/
console.log(box.getArea());

box.getArea.call(box);

var boxes: HasArea[] = [new Square(5), new Rectangle(10, 10)];
boxes.forEach(b => {
  console.log(b.getArea());
});
```

First gets compiled down to typescript like:

```typescript
// Setup
class Rectangle {
  constructor(public width: number, public height: number) {}
}

class __ExtendRectangleBacking
{
	constructor (public Instance: Rectangle)
	{
		this.getArea = __ExtendRectangleBacking.prototype.getArea.bind (this);
	}

	getArea ()
	{
		debugger;
		if (this.Instance["getArea"] !== undefined)
		{
			return (this.Instance["getArea"].apply (this.Instance, Array.prototype.slice.call (arguments)));
		}

		var s = this.Instance;
		return (s.width * s.height);
	}
}

var __ExtendRectangle = function (Target: Rectangle)
{
	return (new __ExtendRectangleBacking (Target));
};
// end setup

// Challenge course begins
interface HasArea {
  getArea(): number;
}

function fn1(x: HasArea)
{
  console.log(x.getArea());
}

fn1(__ExtendRectangle (new Rectangle(10, 10)));

var a: any = Math.random() > 0.5 ? __ExtendRectangle (new Rectangle(3, 3)) : { getArea: function() { return 3; } };
console.log(a.getArea());

var s = new Rectangle(5, 5);
var f = __ExtendRectangle (s).getArea.bind(s);
console.log(f);

class Square {
  constructor(public size: number) { }
  getArea() { return this.size * this.size; }
}

// Worst case:
var box: Square | Rectangle = new Square (137);

console.log(__ExtendRectangle (<Rectangle>/*Probably*/box).getArea());

__ExtendRectangle (<Rectangle>/*Probably*/box).getArea.call(box);

var boxes: (HasArea | Rectangle) [] = [new Square(5), new Rectangle(10, 10)];

boxes.forEach(b => {
  console.log(__ExtendRectangle (<Rectangle>/*Probably*/b).getArea());
});
```

And then emited as javascript like:

```javascript
// Setup
var Rectangle = (function () {
    function Rectangle(width, height) {
        this.width = width;
        this.height = height;
    }
    return Rectangle;
}());
var __ExtendRectangleBacking = (function () {
    function __ExtendRectangleBacking(Instance) {
        this.Instance = Instance;
        this.getArea = __ExtendRectangleBacking.prototype.getArea.bind(this);
    }
    __ExtendRectangleBacking.prototype.getArea = function () {
        debugger;
        if (this.Instance["getArea"] !== undefined) {
            return (this.Instance["getArea"].apply(this.Instance, Array.prototype.slice.call(arguments)));
        }
        var s = this.Instance;
        return (s.width * s.height);
    };
    return __ExtendRectangleBacking;
}());
var __ExtendRectangle = function (Target) {
    return (new __ExtendRectangleBacking(Target));
};
function fn1(x) {
    console.log(x.getArea());
}
fn1(__ExtendRectangle(new Rectangle(10, 10)));
var a = Math.random() > 0.5 ? __ExtendRectangle(new Rectangle(3, 3)) : { getArea: function () { return 3; } };
console.log(a.getArea());
var s = new Rectangle(5, 5);
var f = __ExtendRectangle(s).getArea.bind(s);
console.log(f);
var Square = (function () {
    function Square(size) {
        this.size = size;
    }
    Square.prototype.getArea = function () { return this.size * this.size; };
    return Square;
}());
// Worst case:
var box = new Square(137);
console.log(__ExtendRectangle(box).getArea());
__ExtendRectangle(box).getArea.call(box);
var boxes = [new Square(5), new Rectangle(10, 10)];
boxes.forEach(function (b) {
    console.log(__ExtendRectangle(b).getArea());
});
```

- [ ] Allows use on uncertain types. See: https://github.com/Microsoft/TypeScript/issues/9#issuecomment-236507167
- [ ] Can be used on union types

If the extension method returns s than accessing width gets converted to:

```typescript
box.SomeMethodThatReturnsS ().width
```
Gets converted to:

```typescript
__ExtendRectangle(box).SomeMethodThatReturnsS ().Instance.width
```

Any thoughts?

**Comment by RyanCavanaugh** [comment]
You've "cheated" here
```ts
var a: any = Math.random() > 0.5 ? new Rectangle(3, 3) : { getArea: function() { return 3; } };
console.log(a.getArea());
```
The declared type of `a` is `any`. There's no justification to wrap the call to `new Rectangle` in a `__ExtendRectangle` call here but *not* wrap the same call on the line `var s = new Rectangle(5, 5);`. They're indistinguishable from the compiler's perspective.

The GC implications of this are terrifying - you're causing something like 5 extra allocations *per function call* even in places where 0 should be needed.

**Comment by KingDavid12** [comment]
Okay. On point number one or "cheating" just adjust it so that items only get wrapped when the extension method gets called. I am guessing something like this:

```typescript
// Setup
class Rectangle {
  constructor(public width: number, public height: number) {}
}

class __ExtendRectangleBacking
{
    constructor (public Instance: Rectangle)
    {
        this.getArea = __ExtendRectangleBacking.prototype.getArea.bind (this);
    }

    getArea ()
    {
        if (this.Instance["getArea"] !== undefined)
        {
            return (this.Instance["getArea"].apply (this.Instance, Array.prototype.slice.call (arguments)));
        }

        var s = this.Instance;
        return (s.width * s.height);
    }
}

// Challenge course begins
interface HasArea {
  getArea(): number;
}

function fn1(x: HasArea)
{
  console.log(x.getArea());
}

fn1(new Rectangle(10, 10));

/* Okay. I admit I don't see how to get around this one: */
var a: any = Math.random() > 0.5 ? new Rectangle(3, 3) : { getArea: function() { return 3; } };
console.log(a.getArea());

var s = new Rectangle(5, 5);
var f = __ExtendRectangle (s).getArea.bind(s);
console.log(f);

class Square {
  constructor(public size: number) { }
  getArea() { return this.size * this.size; }
}

// Worst case:
var box: Square | Rectangle = new Square (137);

console.log(__ExtendRectangle (<Rectangle>/*Probably*/box).getArea());

__ExtendRectangle (<Rectangle>/*Probably*/box).getArea.call(box);

var boxes: (HasArea | Rectangle) [] = [new Square(5), new Rectangle(10, 10)];

boxes.forEach(b => {
  console.log(__ExtendRectangle (<Rectangle>/*Probably*/b).getArea());
});
```

Thanks for pointing that out. Ahh well. I was just hoping that extension methods might be possible in TypeScript and If possible suggest a way to do it.

**Comment by SamuelEnglard** [comment]
I think for those that really want extension methods it could be done as an add-in or a preprocessor

**Comment by KingDavid12** [comment]
Do any such things already exist?

**Comment by SamuelEnglard** [comment]
@KingDavid12 in theory #6508 could eventually allows this. In the short term the TS language service could be used to process the syntax tree and do the very transformation suggested above in a step before the compiler runs. If such a system currently exists I don't know, but I vaguely remember people talking about such things in other issues.

**Comment by RyanCavanaugh** [comment]
I mean, if it could be accomplished correctly through a syntactic transform, we'd do it. The problem is that it can't.

**Comment by series0ne** [comment]
Based on this article, [(http://perfectionkills.com/extending-native-builtins/)] extension methods/(properties?) should be syntactic sugar for a "static" object, with the same semantics as C#. This ensures that extensions follow the SOLID rules and don't allow existing objects to be modified

## Example

```
static class StringExtensions {
    public static isUpper(this value: string): boolean {
        ... logic;
    }
}

var myString: string = "Hello World";
var result: boolean = myString.isUpper();
```

## Compiled

```
var StringExtensions = {
    isUpper: function(value) {
        ...logic;
    }
}

var myString = "Hello World";
var result = StringExtensions.isUpper(myString);
```

## Considerations

- extension classes should be open ended
- extension method parameters should work with existing types, type aliases, unions, and interfaces

**Comment by RyanCavanaugh** [comment]
@series0ne your comment suggests you haven't read the thread. This has been brought up and explained why it doesn't work several times

**Comment by electricessence** [comment]
If you want to see the closest real thing in TS/JS of extensions, look at RxJS.
They combine extending with granular importing of functions but it's obviously very controlled and wouldn't work automatically for say ... an ```interface```.

**Comment by JHawkley** [comment]
So, I'm honestly really surprised that this has been talked to death and nothing has been worked out yet.  I'm also surprised this was labeled 'out of scope'.  That is very short-sighted, especially since extensions belong in TypeScript more than they belong in C# for all the same reasons that TypeScript decided to use a structural type-system in the first place.

This is gonna be a long one, and I apologize for that, but I think it has to be long to set up the case properly.  I can see a lot of confusion as to the "what" and "why" for extensions.

### TypeScript and Structural Typing
If I'm going to make a case for extensions, I need to make sure everyone knows what problem they are trying to solve.  Yep, they solve a problem that exists in TypeScript today; they're not just a "nice thing to have".  They facilitate a kind of programming that can be extremely powerful and useful.  I think the best way to do this is to teach everyone the problem that structural typing solved in TypeScript's youth and how that leads to extensions.

When the powers that be started to work on TypeScript, they almost immediately ran into a problem: JavaScript is very promiscuous with the structure of its objects.  In particular, two objects in JavaScript can have entirely different prototypes, but still have the same properties and much of the same behavior.  If these were to bounce their way into TypeScript code through the same entry-point, it would need the equipment to deal with them.

Let's take a look at a very simple kind of structure, the 2D vector:  `{ x: number, y: number }`  Just how many ways can you think of obtaining such a structure in TypeScript?  Here are few!

```TypeScript
// Importing one from a third-party library.
import VectorPro from 'lib/vector-pro';

// Making your own vector class.
class MyVector {
  constructor(public x: number, public y: number) { }
  /* ... other stuff ... */
}

// Coming in from your library's public API.
export function gimmieVectorsNomNom(v: { x: number, y: number }) { /* do noms */ }

// A raw JavaScript object.
let v1 = { x: 0, y: 0 };
```

Now, imagine writing a library where all those types can find their way into your code.  TypeScript solved this problem by allowing you to wrangle them all into an `interface` based on their common structure.

```Typescript
interface VectorLike {
  x: number;
  y: number;
}
```

But you have a new problem.  While you can handle the **structure** of a vector, you can't easily attach **behavior** to it.  But you are a clever object-oriented programmer.  Wrappers are a perfect solution!

```Typescript
class VectorWrapper<T extends VectorLike> implements VectorLike {
  get x(): number { return this.source.x; }
  set x(v: number) { this.source.x = v; }
  get y(): number { return this.source.y; }
  set y(v: number) { this.source.y = v; }
  
  constructor(public source: T) {}
  
  length(): number { /* length stuff */ }
  add(other: { x: number, y: number }): this { /* add stuff, return this */ }
  scale(scalar: number): this { /* scaling stuff, return this */ }
}
```

All done!  Except you're not.  If you're wrapping `VectorLike` objects into this thing, you now have a new problem.  You have to begin unwrapping them whenever you want to send them out of your API, or you can start handing out `VectorWrapper` objects, but you have to start checking to see if incoming `VectorLike` objects are already instances of `VectorWrapper` so you don't wrap them twice!  Hmm...  Not so perfect, after all.

This is when you start thinking arrogantly and make the mandate, "in order to use my library, you WILL use MY library's vector class!"  All this does is shift the complexity of juggling these types to your library's users.  Plus, it adds yet another vector class floating around in the JavaScript run-time.  No one is happy with this.

"But, wait!"  Some clever programmer interjects, "you're going about this all wrong!  Here, look!"

```Typescript
// Just make all your vector math into functions instead!
function length(op: VectorLike): number { /* length stuff */ }
function add(op1: VectorLike, op2: VectorLike): VectorLike { /* add things, return op1 */ }
function scale(op: VectorLike, scalar: number): VectorLike { /* scale stuff, return op1 */ }
```

Wow, look at that!  Behavior that works with `VectorLike` and doesn't need any additional abstraction!  What this clever programmer stumbled upon is something called procedural programming.  But you might be seeing another problem:

```Typescript
import { add, scale } from 'vector-math';

let v1 = { x: 2, y: 4 };
let v2 = { x: 6, y: 8 };
let v3 = { x: 10, y: 12 };
let acc = { x: 0, y: 0 };

// Get the average of three vectors.
let avg = scale(add(add(add(acc, v1), v2), v3), 1 / 3);
```

That's quite the mess!  Not to mention you have to start importing these vector math procedures where ever you want to use them in your code, polluting your scopes with functions floating all over the place.  Unfortunately, in this case our behavior is too far separated from the structures they're made to operate on.  Its uncomfortable and unwieldy, and why we moved away from it long ago.

So, while TypeScript solves the problem of **structure** with interfaces, it is far from graceful when dealing with the **behavior** of these structures.

But wait a minute!  Maybe our procedural programming friend is on to something.  The first parameter of all those vector math functions all share the same structure...

### Extensions are Procedural

If you have a look at the [Wikipedia page on Extension Methods](https://en.wikipedia.org/wiki/Extension_method), you'll see a quote right at the top by the principal developer of C#'s compiler: "Extension methods certainly are not object-oriented."

And he is right, and you can see this in the way C#'s extension methods are declared.  Let's do that with our vector math procedures for comparison, real quick.

```CSharp
public static class VectorMathExt {
  public static Double Length(this IVector op1, IVector op2) { /* length stuff */ }
  public static IVector Add(this IVector op1, IVector op2) { /* add stuff, return op1 */ }
  public static IVector Scale(this IVector op1, double scalar) { /* scale stuff, return op1 */ }
}
```

Now, with these extensions in scope, where ever we have an `IVector` object, we can call these procedural functions.  As extensions, the compiler marries this behavior to any `IVector` making it so we can call `vec.Add(...)` instead of having to use `VectorMathExt.Add(vec, ...)`.  Since this is all handled for us by the compiler, we also only have to do a single import statement, and since they're namespaced into `VectorMathExt`, the scope pollution is minimal.  But, once they are in-scope, the compiler handles the rest, replacing our nice dot-operator calls to the extension methods with the more unwieldy static class calls.

You might have noticed a while ago I keep bolding a couple of words when they're near each other: **structure** and **behavior**.  That's because, deep down, this is how programs are divided up.  You cannot have one without the other and still have a working program.  There are many ideas on how these two concepts should inter-relate.  Object-oriented programming has the structure owning the behavior: the familiar class.  Functional programming has the behavior owning the structure: data lives in the stack (as much as possible).

In between, you have procedural programming, where structure and behavior do not own each other; they stay separate.  Instead, structure and behavior infer each other.  A procedural programmer's job is to find structures and behaviors that agree with each other and use them together, but, as we saw, this is very tedious.  Now, if there were a way to map procedural functions on to compatible data structures automatically, this would be a great relief for procedural programming in JavaScript.

Basically, extensions take procedural programming and give it that object-oriented flavor we crave, and it does all this without destructively changing the underlying types.  That C# `IVector` interface is the same as our TypeScript `VectorLike` interface, just an `x` property and a `y` property, with or without extension methods.

But many of you already know **how** extensions work.  It's the fact that this is compiler magic and these extension methods cannot be laid upon the JavaScript world neatly.  These extension methods vanish when they're not used in a method call!  They won't act like they have these behaviors if they're passed into different scopes that lack these extensions!  It would be weird!

### It's Already Weird, Dude
The idea of call-site rewriting seems strange to people coming out of JavaScript, and that's largely because they're coming from an interpreted language; JavaScript doesn't have a compiler.  To a JavaScript developer, there exists only 'run-time' and anything else is sorcery and witchcraft.

However, just by virtue of using TypeScript, you must deal with 'compile-time' abstractions.  I would bet that everyone here stumbled on `someVariable instanceof MyInterface` at one point or another and thought for a moment that TypeScript's compiler had a bug in it.  But you figured it out; you understand now that 'types' are a compile-time thing and 'values' are a run-time thing, and there is a division between them.

And that is what people will learn to understand with extensions.  Extensions are another one of those compile-time abstractions that don't exist in the same way in run-time.

### But Only if You're Ready
Keeping that in mind, it would be a terrible idea to just start piling extensions onto everything.  C# made the smart choice of making extensions opt-in; in order to use them, you must bring them into scope with the `using` directive.  Other languages which have features similar to extensions, like Rust's traits, are also opt-in and strict about it.  Being opt-in, we also get the advantage that adding extensions to the compiler will not break existing code.

It should be little different for TypeScript, especially if it will take some time for programmers to adapt to the concept.

### The Missing Half
TypeScript provides a means of conjuring structure to help cope with that structural type-system, the `interface` and `type` keywords.  Unfortunately, it has no simple means of joining behavior to these structures.  This is what extensions provide and the problem they solve in TypeScript.

Ponder the following for a moment:
**Interfaces describe the structure of a type.**
**Extensions describe the behavior of a structure.**

It's starting to sound a lot like extensions are the missing compliment to interfaces.  Interfaces cannot have behavior; they can define **space** for behavior, but that isn't the same thing as having it.  Where-as extensions cannot have structure, but they can define what particular structures are capable of doing.  It's getting harder and harder to see how you can have one of the features but not the other.

### Conclusion
As I see it, we have a feature that is only half-implemented in TypeScript.  We got interfaces that poof structure into existence; where is the thing that poofs behavior into existence to go with it?  It's hard to argue that extensions do not strongly compliment TypeScript's structural type-system.  It would make coping with many of the challenges that exist in the domain of JavaScript so much easier.

That being said, I have been working on a complete proposal, but I think I'll wait to see if the mood changes, after everyone has digested this, before I finish it.

**Comment by RyanCavanaugh** [comment]
@JHawkley please don't confuse the issue - it's not that extension methods aren't awesome and desirable. Everyone who uses C# recognizes their value. It's that they *cannot*[1] (**cannot** [2]) be added to the language in a way that actually works, for any suitable definition of "works". See https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592

[1]: SRSLY CANNOT
[2]: can't even

**Comment by SaschaNaz** [comment]
@JHawkley If your proposal really fits in a JS world then maybe you can post it on [esdiscuss](https://esdiscuss.org) or [WICG](https://discourse.wicg.io/c/js) so that the whole JS community can benefit from it :D

**Comment by JHawkley** [comment]
@RyanCavanaugh I actually kind of found your challenge course kind of funny, at least how you began it.  Thing is: C# can't run that course with its extensions, and it **is** C# and it **does** have the CLR.

In C#, having an extension mapped to a type does not actually change the underlying type.  If an object didn't match an interface before it had an extension, it still won't after.  The compiler would helpfully inform you of the type mismatch.  Extensions (at least how C# implements them) do not change types; they only associate free-floating procedures to types, making calling those procedures more convenient.

Fundamentally, your challenge course is confusing two concepts: extensions vs monkey-patches.

The core difference is one virtually extends an object in **compile-time** and the other ACTUALLY extends it in **run-time**.  Your challenge course is making the assumption that extensions would need to work the same way in both domains, and if they don't, they fail and are therefore unworkable as a feature.

But extensions are not monkey-patches.  There are other proposals for this kind of feature elsewhere.  Extensions solve a different problem from monkey-patches, and so must be given different treatment and have different expectations.

They do work in TypeScript, just not in the same way as a monkey-patch.  Also, the extensions that TypeScript would produce would also work fine in JavaScript; the popular library [UnderscoreJS](http://underscorejs.org/) is an example of what a TypeScript library built on extensions would compile into.  Perfectly usable!  (In fact, a prime candidate for having its definition files re-written to support extensions.)

_Also, as an aside:_ the only language I currently know of that could actually run your challenge course is Scala, with its implicits.  Those are, uhh...  really wild, though.  [ScalaJS](https://www.scala-js.org/) exists and it does manage to bring the feature to JavaScript, but I am certain that the structures it compiles into are probably unsuitable for use in bare JavaScript, though.

**Comment by JHawkley** [comment]
Actually, @RyanCavanaugh, I think I see what you were saying.  Like a lot of people, I got caught up in the whole idea of C#-style extension methods when your suggestion clearly states "adding members to an existing type's prototype"; that's obviously a case for run-time monkey-patching.  And that definitely wouldn't work, especially now that we have union and intersection types to deal with.  Although compile-time abstractions were discussed thoroughly, it wasn't really what the suggestion was originally about.

But, now I'm not sure what to do.  I still think compile-time abstractions for associating behavior to types does work for TypeScript and has great merit for both it and JavaScript.  It feels like I should create a new issue (or find an old one) that pertains specifically to compile-time extension methods, but I am pretty certain that, by now, the issue would be immediately marked a duplicate and I'd be sent right back here.  I mean, this IS issue number 9.  It's an old, old discussion...

But, I've expended considerable energy on this already, and I think I laid the case for compile-time extension methods as well as I could, where it was mostly discussed.  At this point, either people understand the difference and see that it works or the idea dies here.  I hope it doesn't, but TypeScript will continue to be a great language all the same.

**Comment by RyanCavanaugh** [comment]
Again, *it would be awesome* to have that feature. It's super handy. If we could wave a magic wand ✨ and make extension methods happen, we most likely would. There's no question about the desirability of the feature.

The constraint of compiling to simple JS simply makes it impossible to implement. If someone somehow came up with a way to introduce that behavior (in terms of JS emit) while still meeting our other design constraints, I think it'd be something we were really excited about. But the fact that five years later no one has come up with anything, despite everyone wanting it and wishing it existed, strongly implies that it cannot be done.

**Comment by danfma** [comment]
I really agree with you, @RyanCavanaugh!

**Comment by SaschaNaz** [comment]
I think something like [binding+pipelining](https://github.com/tc39/proposal-bind-operator/issues/37) will work without the typing problems because the syntax is explicit.

**Comment by electricessence** [comment]
@RyanCavanaugh I still don't understand why it can't be simple syntactic sugar.
All extensions are simply a static method/function that takes this as the first property.
Why can't TSC simply rewrite the function insanity for us?

```ts
interface IEnumerable<T> {
   getEnumerator():IEnumerator<T>
}
```
```ts
// Extensions:
export function where<T>(this a:IEnumerable<T>, p:Predicate<T> ):IEnumerable<T>
{ /*... */ }
export function select<T,TResult>(this a:IEnumerable<T>, s:Selector<T,TResult>):IEnumerable<TResult>
{ /*... */ }
```
```ts
import {* as Linq} from "Linq";

function filter<T,TResult>(source:IEnumerable<T>, p:Predicate<T>, s:Selector<T>)
{
    return source.where(p).select(s);
}
```
... results in:
```js
function filter(source, p, s)
{
    return select(where(source,p),s);
}
```

As always, I understand that there could be collisions if the existing objects had a ```.where``` or ```.select``` member.  But: 1) In this case, the intention is clear because of the restrictive interface.  2) The compiler should be able to detect the collision and warn for both the cases where the extension itself is breaking the rule, or being broken when applied to the object being extended.  3) Also, the compiler should be able to identify if there is a difference in signatures between the two and even potentially create a virtual overload because of it.

**Comment by electricessence** [comment]
@RyanCavanaugh And again, I understand that implementing extensions is probably going to break things in JS but it could be done in a restrictive way first pass where everything has to be in order for it to work right.

**Comment by RyanCavanaugh** [comment]
Here is a non-exhaustive list of problems with this approach, as already described at https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592

1. What happens when an extension method is invoked on an expression of type `any` ? We'll have no idea which method to rewrite to, let alone whether we should at all. A slight change to a .d.ts file that you don't even own might cause an expression to become `any` *and break your program* - super bad. How would you even diagnose this? It just looks like a super bad compiler bug.
2. What happens if you simply reference an extension method, rather than invoke it? The best-case scenario would be a `bind` invocation, but even this isn't the same (reference identity stops working, for example). Compound this with the previous bullet point.
3. What happens if you have a union of two objects and two extension methods extending both types in the same way? Or if you have a union where one type has the method built-in and the other has the method from an extension. By all accounts you would expect to be able to invoke the method, but we'd have to somehow generate code that *at runtime* would distinguish the two objects and pick which method to invoke. In the general case, this is simply impossible. Compound this with the previous two bullet points.
4. What happens if you e.g. add `foo` as an extension method to `bar` and pass a `bar` to a function that expects ` { foo(): string }` ? We'd *have* to disallow this because we can't just go rewrite the inside of the function. But now it's completely opaque why *you* can invoke a method, but you can't pass it to a function that wants to invoke the method on your behalf. It just looks like a compiler bug.

**Comment by RyanCavanaugh** [comment]
5. What about intersection types? What should happen for a type `A & B` that has an extension method `foo` for both `A` and `B` with the same name, but different implementations for `A.foo` and `B.foo`? I guess you just can't invoke it, but that breaks the contract for intersection types -- now `A & B` is no longer a substitute for `A` nor `B`. Technically we would just pick the first method we found and invoke it, but now your program's behavior would be dependent on the compiler's internal ordering of the constituents of intersection types, which is not at all specified (would the type `B & A` now need to be *distinct* from `A & B` ?)
6. We would be strictly unable to ever change anything in the type system again without fear of breaking someone's program's behavior. Overload resolution, for example -- we really want to tweak the rules here, but now returning an `A` instead of a `B` due to resolution changes might result in a different method being invoked at runtime and cause a previously-working program to fail (or worse, corrupt data)

**Comment by electricessence** [comment]
@RyanCavanaugh all valid points that I'm aware of.  Hence why I'm suggesting it may require a heavy amount of restrictions to work.  Compile time collision detection would be essential, and not optional.

- Import two extension libs that directly collide because their signatures match/overlap?  Compiler disallows.
- Using ```any``` as the type in an extension's ```this``` parameter?  Compiler disallows.
- Applying an extension to a value that is of ```any``` could be warned against or simply ignored.  Yes, this might look funny.   But I would suggest a compile warning by default if an extension is imported and the ```any``` value has it applied.  Again, extensions would be a 'special' feature.  You'd need to then have a compiler flag to say "ignoreExtensionAny".
- At first you could be very explicit with extensions that it has to be a perfect match or they won't be applied.  So any unions would fail.  Intersections for the most part shouldn't, but I get that intersections are a bit funny too.

**Comment by RyanCavanaugh** [comment]
7. Basic JS premises like `bar["foo"] !== undefined` implying `"foo" in bar` are now broken, unless we would want to somehow rewrite the `in` operator in certain cases (how??). It's not clear how you could do this in a way that doesn't fundamentally change the runtime behavior of JS in many cases.
8. Basic JS premises like `const x = "foo"; bar[x]();` being identical to `bar.foo();` are now broken, unless we want to start doing aggressive constant folding, but even that wouldn't be sufficient. It's hard to imagine keeping a straight face on the "all JS is TS" promise here.

**Comment by RyanCavanaugh** [comment]
> Applying an extension to a value that is of any could be warned against or simply ignored. 

This is circular logic. How do you know that an invocation of a method of an `any` expression is an extension method call rather than a normal call? You don't, because it's `any`.

**Comment by electricessence** [comment]
Yeah, like # 8 is great example.  But I'm suggesting that those are not cases where extensions be valid.
It would have to be a function, and it would have to be dot notation.

**Comment by electricessence** [comment]
| You don't, because it's any.
```any``` does throw a wrench in the gears.
Maybe the use of ```any``` is somehow disallowed in modules that use extensions.

**Comment by electricessence** [comment]
I guess ultimately, I don't see it as impossible.  It's just problematic or difficult to get right.

**Comment by RyanCavanaugh** [comment]
I mean, it's *possible* as long as you're willing to accept a feature that is outright dangerous to use, has non-working scenarios that look like compiler bugs, restricts the use of many common JS idioms, requires type-directed emit, and doesn't interop with plain JS at all.

It's also possible to wire your house with uninsulated 36 gauge wire and bubble gum as long as you're willing to accept it sometimes catching fire. :wink:

**Comment by series0ne** [comment]
@RyanCavanaugh could you provide a sort of similarity in terms of how dangerous extension methods are in C# vs TS? - I mean we're all talking about extension methods, essentially because they're in C#, and we'd like to see them in TS...

...brb, need to go rewire my house quickly...I can smell burning gum!

**Comment by electricessence** [comment]
@RyanCavanaugh at your last comment.
I have to respect consistency.  So I agree, adding a feature that creates confusion is bad.
Thank you for the in-depth reply.

**Comment by electricessence** [comment]
@series0ne if I may help out here and summarize...
Just the notion of ```any``` is enough to create a serious issue.
There's just too much opportunity for collision.  And as I chatted with Ryan about, you could solve these headaches, but it would create different headaches and inconsistencies that make it not worth it.

**Comment by RyanCavanaugh** [comment]
There's no danger in C# because C# a) resolves all call sites at compile-time, b) uses a nominal type system, c) uses an explicit opt-in mechanism to provide dynamic method calls (in which extension methods simply don't work), and d) doesn't allow method references in typeless contexts. Consider this C# code:
```ts
            IEnumerable<int> enumer = null;
            List<int> list = null;
            Func<int> x = enumer.Count;
            int y = list.Count;
```
The IL for the last two statements there is *very* different!

We also have to consider code like this
```ts
            var z1 = enumer.Count; // Error
            var z2 = list.Select; // Error
```
C# rightly errors on these, but TS really *can't* - that's valid JS and we have to emit something.

**Comment by electricessence** [comment]
@RyanCavanaugh, maybe you've already done so, but I think this topic is worthy of a blog post. :P
"Why JS/TS cannot have extensions"

**Comment by series0ne** [comment]
Essentially it seems then, that we will simply have to stick to statically bound method calls in TypeScript, the same way we would have done it with C# before extension methods :(

**Comment by RyanCavanaugh** [comment]
The bind operator https://github.com/tc39/proposal-bind-operator is very promising and would accomplish nearly all the extension method scenarios quite cleanly. So keep some hope :wink:

**Comment by series0ne** [comment]
😯 @RyanCavanaugh you would not believe the post I'm writing right now...which, without seeing the bind operator proposal, is virtually identical (great minds...?) here goes...

Essentially the only way I can see this working is to introduce an operator which instructs the TypeScript compiler, _this is an extension method call_! This borrows from C++

**Example**

```
class StringExtensions {
    public static isUpper(this string value): boolean {
        return value.toUpperCase() === value;
    }
}

var x: string = "Hello World";

x->isUpper();

//or

x::isUpper();
```

Even if `isUpper `was already a member of `String.prototype`, TypeScript would know to use the extension method, not the prototype bound method, unless the operator changed.

In terms of dealing with things like union or intersection types, I think the only solution would be to make the extension method **invariant**; for example

```
var x: string | number = 0;

x::isUpper(); // Error, cannot call this because the extension method expects type "string"
```

Thus, you would require an overload to allow this behavior.

Again, I did **NOT** read about the bind operator before writing this, and it's just escaped my noodle, so there are likely some gaping holes!

**Comment by JHawkley** [comment]
Yeah, the bind operator would go a long way toward giving us extension-like behavior.  It still isn't monkey-patching, won't fix types like you might want, and has its own challenges for integrating into TypeScript, but not impossible challenges.

TypeScript would have to start tracking the type of `this` in a function more aggressively, for instance.  We already have that ability to explicitly specify its type with the fake `this` parameter, but it would have to be expanded to automatically infer the type of `this` for functions pulled from class prototypes and the like.

Anyways, I still think TypeScript can benefit from a compile-time abstraction for associating behavior to types, especially since it has one for structure.  Many current JavaScript libraries are created in a manner that would work well with it and they wouldn't be helped by the bind operator.  Also, it would reduce the compulsion to duplicate function implementations in separate classes or use unintuitive forms of inheritance, all just to get the nicer feeling method-call syntax.  I've done that before...

I think I may submit my current proposal under the terminology the programming language D uses, ['Uniform Function Call Syntax'](https://en.wikipedia.org/wiki/Uniform_Function_Call_Syntax), and try to keep the use of the words 'extension' or 'extends' to a minimum.

**Comment by danfma** [comment]
Then maybe the Scala extensions would be help this kind of problem. In Scala, I can define a class which will provide methods and properties to an instance. So, in resume:

```typescript
extension class MyExtension(target: any)
{
    printHello() {
        console.log("DO MY CODE");
    }
}

var something: any = ...;

something.printHello();
```

could be transformed to this equivalent:

```typescript
class MyExtension(target: any)
{
    printHello() {
        console.log("DO MY CODE");
    }
}

var something: any = ...;
var something$MyExtension = new MyExtension(something);

something$MyExtension.printHello();
```

and then to the equivalent javascript code. Thus, In the end:

* We won't have any problem with name collision because to use an extension I need to import it;
* All the transformation is done by the compiler;
* Yes, we have problems with additional allocation of objects when creating a new extension, but that is a feature that can or can't be used, so we just need to used it wisely;

**Comment by RyanCavanaugh** [comment]
@danfma I don't see how that's meaningfully different from the dozens of upthread suggestions for similar rewriting of call sites, nor how it addresses the serious problems with that approach.

**Comment by danfma** [comment]
@RyanCavanaugh, that just another suggestion like many others above, but let me try to explain some points:

* Maybe I'm wrong but you guys ignored the extension function as a global extension because of global variable pollution in javascript. With this example, we don't have that.
* We can import only the extensions that we want to apply (which could be anyone above since simple functions to extension classes), the only problem with this will be when two or more extension collide in name, but you could resolve that by applying an alias at the import and because each extension is dependent of an extension object which the compiler will create at the usage, then you won't have the problem of two extensions with the same name because they are on different objects.

**Comment by danfma** [comment]
About the extension classes, it has the advantage to allow you to extend with more than one method, but it will make the compiler life hard when having two extension classes with the same function name. One solution for that, is to not resolve automatically when that occurs, so if a name collides, the compiler can raise an error, and let the user resolve, by manually creating the extension by hand, for example.

**Comment by RyanCavanaugh** [comment]
All of the problems raised at https://github.com/Microsoft/TypeScript/issues/9#issuecomment-74302592 and https://github.com/Microsoft/TypeScript/issues/9#issuecomment-263978824 apply

**Comment by danfma** [comment]
ok, I will check!

**Comment by danfma** [comment]
@RyanCavanaugh, I got your points, but I still disagree in some points. Javascript itself is a dangerous language because you can do a lot of things to bug your own code and like the TypeScript team always  talks 'TypeScript is JavaScript', so it inherits all its problems. 

So when you say `It's also possible to wire your house with uninsulated 36 gauge wire and bubble gum as long as you're willing to accept it sometimes catching fire`, you are right, but you already accepted that too! :D 

What I want to say is that even in C# you can have problems by using `dynamic` or Reflection, so I don't see the difference here. 

Anyway, I will stop here and thank you for the long list provided before, I didn't see that.

**Comment by AndrewLang** [comment]
👍

**Comment by co3carbonate** [comment]
An alternative is to include a forward declaration of the methods you are extending in the class, then modifying the prototype to extend them. For example:

```
class Greeter {
    greeting: string;
    say: Function;
    constructor(message: string) {
        this.greeting = message;
    }
    greet() {
        return "Hello, " + this.greeting;
    }
}

Greeter.prototype.say = function(msg:string) {
    alert(msg);
}

let greeter = new Greeter("world");
greeter.say(greeter.greet()); // alerts "Hello, world"
```

Hope it helps

**Comment by mikeaustin** [comment]
Honestly, I'd love to see a move away from internal methods at all... some day? Monkey-patching and static helpers only go so far. CLOS and Dylan are good examples of adding methods to classes at any time -- without monkey-patching, and they're lexically scoped. The problem with C# extension methods is that they are statically dispatched, just some syntactic sugar.

If every method was external, you could do something like this:

var _capitalize = capitalize || new Map();

String.addMethod(_capitalize, function() {
  return this.slice(0, 1).toUpperCase() + this.slice(1);
});

"foo".invoke(_capitalize);

// Implementation

Object.prototype.addMethod = function(method, func) {
  method.set(this, func);
}

Object.prototype.invoke = function(method, args) {
  return method.get(this.constructor).apply(this);
};

// Syntax thoughts, where extend creates a locally scoped external function.

extend String {
  capitalize() { ... }
}

"foo".capitalize();


Just some thoughts :)
Mike

**Comment by profound7** [comment]
Some of the examples @RyanCavanaugh wrote on why rewriting call-sites doesn't wouldn't work for JS targets, actually works in Haxe (similar to TypeScript and can transpile to JS and others).

[See here for one of the examples, in Haxe](http://try.haxe.org/#53c8f). Click the build button to see the trace outputs and the JS transpiled output.

Some of his examples don't work, but the Haxe compiler actually manages to catch some of those errors ("x has no method getArea"). However, using the any type (Dynamic in Haxe), will definitely not work, but haxe coders know why they don't work, because in haxe docs, it was stated that we should avoid using Dynamic as it turns off certain type checking features.

To make most of Ryan's examples work in haxe (haven't tried all of them), simply type cast or add type hinting so the compiler knows what to do. Of course, if someone convinced the compiler that a chicken is a duck, but it doesn't quack at runtime, then the problem isn't that the feature is impossible. It is the person who convinced the compiler that the chicken is a duck.

For union types (Square and Rectangle; where one has instance method getArea and the other a static extension getArea), the compiler should show an error because getArea does not really exist for one of them. It should be made that once the type is checked (through an if guard or something), then getArea should work (since within the if statement, the compiler can tell if its a square or rectangle).

Static extension methods can only work if the compiler is able to know the type as it is a compile-time feature. Programmers shouldn't expect something untyped/not properly typed/unverified to take advantage of a compile-time feature that requires type to be known. I think it is possible to implement and that it is also reasonable to expect it not to work if the compiler is unable to infer the type.

**Comment by RyanCavanaugh** [comment]
@profound7 Consider this code (Haxe):
```ts
class BetterSquare extends Square {
    public function getArea() {
        return this.x * this.y;
    }
}


class SquareExtension
{
    public static function getArea(square:Square)
    {
        trace("getArea!");
    	return square.x * square.y;
    }
}

class AThing {
    public static function doSomething() {
        var qq: Square = new BetterSquare(10, 10);
        qq.getArea();
    }
}
```
This results in calling the extension method, rather than the `BetterSquare` method. **This breaks JS runtime semantics preservation** -- `qq.getArea()` *must* call the instance method on the class because that's what JS does. The TypeScript type system *never* changes the runtime behavior of code; that is a core principle and a primary guarantee of our language.

**Comment by RyanCavanaugh** [comment]
(Be aware that any for code you can possibly think about emitting, I can show an example of how to use that behavior to break JS semantics)

**Comment by mikeaustin** [comment]
Ok, I get it. It mimics C# in a sense that it prioritizes class methods over extension methods. It seems possible to check if the method exists first, but I assume that would incur some overhead.

Personally, I like the idea of being able to shadow class methods. It opens the door to adding functionality, applying bug fixes, logging, security -- AOP cut-points, etc. If they are lexically scoped, there's no risk to other code (other than trojan horses, I guess). Hmm, now I've got to think :)

I've been toying with the idea lately and came up with this form for extension methods. Thoughts?

https://structura.quora.com/Extension-Methods-in-JavaScript

TL;DR:

```
var _capitalize = Impulse.extend(String, _capitalize, function() {
  return this[0].toUpperCase() + this.slice(1);
});

console.log(_capitalize.apply("foo"));
```

It takes advantage of hoisting and passing the previous method when defining a new one. Applying the method works just like regular functions, but it looks at the type of 'this' and dispatches accordingly. I know there are libraries for multi-methods and stuff, but I wanted to keep it simple.

I use an underscore to not pollute the namespace. lisp-1 vs lisp-2 namespacing debate :)

**Comment by RyanCavanaugh** [comment]
@mikeaustin it's fine for primitives and classes, but with an erased type system there's no prototype to key off of. It's also not really an "extension method" in the sense of something that *appears* as a new method of a given type.

**Comment by series0ne** [comment]
I may have mentioned this before on another thread but I think this problem can only be solved with the introduction of a new operator

Calling an existing method on a type

`obj.foo();`

Calling an extension method on a type

`obj::foo();`

The latter would translate to...

`Extensions.foo(obj);`

This allows a distinct differentiation between prototype bound method and extension method calls, without any ambiguity over which is being called in a particular context.

I know there is an ESNext proposal for a bind operator `::` so not sure how this would impact this suggestion...would always go with C++'s `->` operator if they conflict

**Comment by mikeaustin** [comment]
Hmm, the universal access principle comes to mind when I see two different ways to call a method. Isn't the purpose of extension methods to hide the implementation?

As to Ryan's message, "extension method" isn't really a formal term outside of C# (I think?) I guess I could call what I'm describing as "method shadowing", which is exactly what it's doing.

**Comment by aluanhaddad** [comment]
Scala implicits are sometimes colloquy referred to as extension methods.
I think the term is fairly well understood.

I agree that using a different token breaks the abstraction.
In spite in spite of the semantics being static in C# the syntactic attraction is precisely the point.

One of the problems I have with the `::`  is that it doesn't offer any syntactic abstraction over any existing notation.

I feel like it fails at its purpose, which I think is at least in part the very worthy goal of making own versus ambient desirably indistinct, before it even starts.

It would be great to have extension methods but I can't think of a reasonable way.

**Comment by mikeaustin** [comment]
@aluanhaddad Ahh yes, Scala implicits. Ok, I'll refer to the other way round as "lexically scoped methods". I've been playing a bit more with seamless extension methods in a language I'm prototyping. I've found this works for calling the object method before any extension methods:

`(10).toString ? (10).toString() : _toString.apply(10)`

And the other way around, lexically scoped methods with fallback to object methods:

`typeof _toString === "undefined" ? (10).toString() : _toString.apply(10)`

The language does not do static type-checking, so a TypeScript version would probably need a different approach. The above can be / is slow since it has to do that at each call site. But like I said, I'm just prototyping and getting a sense of how it could work.

**Comment by aluanhaddad** [comment]
In both C# and Scala an instance method will always win out in terms of overload resolution so I think that's a good approach.

The problem with applying that approach in TypeScript is not static checking but rather that it is impossible to do it without resorting to type directed emit, which is not an option.

**Comment by mikeaustin** [comment]
I see the type-directed-emit topic comes up with async/awat, but other than that is it not used anywhere else? Sorry, I don't mean to keep this topic alive forever :)

Side note -- taking your extension methods comments to heart, I've come up with this for my language:

`(foo.toString || _toString).apply(foo, []); // Syntax emmited for: foo.toString()`

Again, it's not going to be fast, but at least it's short.

**Comment by aluanhaddad** [comment]
@mikeaustin I like it, that's a very clean emit, I'm curious to see how your language progresses.

With respect to type directed in async methods, IIRC it only rears its ugly head when you don't use type inference for return types and don't install a global Promise polyfill, as it only considers the type when provided, defaulting back to the global `Promise`.

I couldn't care less about the extra methods available on Bluebird Promises and I never specify the return type of `async` methods.

So I believe the issue is easily dodged, but I may be wrong and it may be affecting me in some other way that I'm not aware of.

**Comment by mikeaustin** [comment]
Sorry for spamming the TypeScript list, but I made a slight change in my extension methods to support traits/mixins that might be interesting. When extension methods were in local scope, there was no good way (aka not using eval()) to dynamically add methods.

So, I encapsulated them in a _methods lexically scoped object:

```
var Iterable = new Trait({
  // iterator() can be an extension method or a plain old function
  map: (iterator) => (func) => { var iter = iterator.apply(this); ... }
});

var _methods = Extension.extend(_methods, String, {
  iterator: () => { return new Iterator(this); }
});

// Bind all the trait methods and pass them to extend String
var _methods = Extension.extend(_methods, String, Iterable.bind(_methods.iterator));

_methods.map.apply("abc", [c => c.charCodeAt(0)]); // [97, 98, 99]
```

I can probably simplify extend() a bit more. The language this is targetting: [Impulse-JS](https://docs.google.com/document/d/1LqEP5ERAjBa1b0iryn88pessVohDyVAPIGgZvMK8IC4/edit?usp=sharing).

**Comment by atrauzzi** [comment]
With the amount of library authoring going on and what seems like increased adoption of TS, now seems like a great time to dust this idea off and see if it's worth dedicating resources to!

Again, biggest thing is for libraries.  I already find myself very badly wishing for some kind of idiomatic way to sprinkle some extensions around.

**Comment by RyanCavanaugh** [comment]
Seems people aren't reading the thread.

**Comment by reddaly** [comment]
I would personally be happy if Typescript had extension methods that used a distinct call syntax vs. normal method invocations.

For instances

```typescript
function Date#toProto(): Timestamp { ... }

function usage() {
  const proto: Timestamp = new Date()#toProto(); // instead of toDateProto(new Date())
}
```

This would allow 
1. call chaining and fluent usage of extension functions. You can write `foo.toBar().toBaz()` instead of `barToBaz(fooToBar(foo))`.
2. easy discovery of extension functions through IDE autocomplete. `new Date()#` or `new Date().` would show `toProto()` in the autocomplete list; selecting that item could auto import the extension function.

I suspect it would avoid some of the complications from https://github.com/microsoft/TypeScript/issues/9#issuecomment-74302592. Making extension functions callable from JavaScript might require some attention, however.

**Comment by RyanCavanaugh** [comment]
@reddaly out of scope for TypeScript. See https://github.com/tc39/proposal-bind-operator though

**Comment by gonzojive** [comment]
What makes the https://github.com/microsoft/TypeScript/issues/9#issuecomment-2229040453 suggestion out of scope for TypeScript?

Extension methods seem like they should be resolved at compile-time based on typing information that's out of scope for ECMAScript. I'm not sure how ECMAScript could possibly provide the same level of utility through the bind operator.

```typescript
function Foo.toProto(): FooProto { ... }
function Bar.toProto(): BarProto { ... }

(new Foo()).toProto() // resolves to first method
(new Bar()).toProto() // resolves to second method
```

**Comment by RyanCavanaugh** [comment]
> Extension methods seem like they should be resolved at compile-time based on typing information

That's type-directed emit. See https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals#non-goals point 4:

> [Do not] Add or rely on run-time type information in programs, or emit different code based on the results of the type system. Instead, encourage programming patterns that do not require run-time metadata.

**Comment by gonzojive** [comment]
Thanks. It seems the more active proposal is https://github.com/tc39/proposal-extensions. I [asked here](https://github.com/tc39/proposal-extensions/issues/15) about the status of that proposal advancing to stage 2. Presumably TypeScript is not interested in introducing its own bind operator before there is consensus to add the operator to JavaScript.

**Comment by Farenheith** [comment]
Hi,

I'd really appreciate some guidance.
I was playing around with ast transformers to see if I could have a custom extension method implementation (here: https://github.com/codibre/extension-methods/pull/3).
I achieve making something like that work:

The extension declaration:
```ts
import { ExtensionMethod } from 'extension-method';

export class ExtensionTest {
    @ExtensionMethod
    static myTest(this: string) {
        return 123;
    }
}

declare global {
  interface String {
    myTest(): number;
  }
}
```

Where it is used:
```ts
import './extension';

console.log("abc".myTest());
```

In the test I'm running, it compiles to something like this:

```ts
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const extension_1 = __importDefault(require("./extension"));
console.log(extension_1.ExtensionTest.myTest("abc"));
```

Which nails the idea, although there are many scenarios I need to explore. 
The issue with this solution, though, is that the extension declared to String here is visible regardless of whether I've imported the extension file. Can I write the "type augmentation" somehow so the extension method will only be visible when I import the file where it is declared?

I see that the extension will only work if I import the extension file, which is ideal for the concept. However, if I could somehow have some "scoped argumentation" for existing types, I'd be able to reflect this behavior also in typescript, but I'm unaware of such a feature.
